### ISSUE TITLE
CDK Timers

### ISSUE THREAD

**Issue by dansteren** [demergent-labs/azle#809]
Implement the [Timers API](https://docs.rs/ic-cdk/latest/ic_cdk/timer/index.html).

> **Warning**
> It sounds like you need to enable the timers in the crate [like this](https://github.com/open-ic/open-chat/blob/master/backend/canisters/group/impl/Cargo.toml#L25).

Resources:

- Forum Post: https://forum.dfinity.org/t/heartbeat-improvements-timers-community-consideration/14201/61?u=dansteren
- Interface Spec: https://github.com/dfinity/interface-spec/blob/master/spec/index.adoc#global-timer

**Comment by dansteren** [comment]
We're currently blocked on this issue because of a crazy compilation issue. When we enable the timers feature in the CDK there is a conflict with boa_engine somehow. If we take either out of the project the compilation succeeds. But when trying to run `cargo test` with both dependencies then we get the error documented in [this gist](https://gist.github.com/dansteren/08372a0192e47639a210a340f6221747).

I've created a reproducible example (on my machine) at https://github.com/dansteren/timers-error.

So for now we are blocked on implementing this until we can get this working.

**Comment by dansteren** [comment]
Jordan tracked the issue down to https://github.com/demergent-labs/boa/commit/3aea26e5419fd67327541b2d4d1323e5e60c2a52#diff-12d0f923375a1d1c03d18a72f1a2777986f1ebd68dd7c7376487865e5865e10eR1481. We just need to make sure we only ever call ic_cdk inside the `wasm32-uknown-unknown` target. So we should be unblocked here once that fix is out.

**Comment by dansteren** [comment]
Spent the day diving into timers. We're hitting some limitations of what we can pass to a callback that needs a static lifetime. Passing data into this closure doesn't work because the references go out of scope.

I've got a very primitve work-around that would require us to look for all calls to `ic.set_timer` and generate static functions for them. This probably won't fly though because analyzing all possible call locations could get very complex.

We will most likely need to build out some sort of mutex to control access to the boa context.

**Comment by dansteren** [comment]
Both `set_timer` and `clear_timer` are implemented on branch [809_timers](https://github.com/demergent-labs/azle/tree/809_timers).

I still need to:

- [ ] implement `set_timer_interval`
- [ ] update tests for set_timer_interval
- [ ] update the documentation (README)
- [ ] repeat for Kybra

**Comment by dansteren** [comment]
> One more thing we probably need to do once async/await is merged is to ensure that the callbacks can be asynchronous

It looks like we will need to wrap the function call in a `ic_cdk::spawn` call. See: https://forum.dfinity.org/t/heartbeat-improvements-timers-community-consideration/14201/85
